### PR TITLE
Add parallel function calling

### PR DESCRIPTION
**Pull request by royziv11**
## Summary

The goal of this PR is to do the following:

- Update this cookbook with the newest schema changes that utilizes 'tools' and 'tool_call' instead of the deprecated 'function' and 'function_call'
- Add a section on parallel function calling

## Motivation

Why are these changes necessary? How do they improve the cookbook?

These changes capture some of our most recent features from Dev Day

---

## For new content

When contributing new content, read through our [contribution guidelines](https://github.com/openai/openai-cookbook/blob/main/CONTRIBUTING.md), and mark the following action items as completed:

- [ ] I have added a new entry in [registry.yaml](/registry.yaml) (and, optionally, in [authors.yaml](/authors.yaml)) so that my content renders on the cookbook website.
- [ ] I have conducted a self-review of my content based on the [contribution guidelines](https://github.com/openai/openai-cookbook/blob/main/CONTRIBUTING.md#rubric):
  - [ ] Relevance: This content is related to building with OpenAI technologies and is useful to others.
  - [ ] Uniqueness: I have searched for related examples in the OpenAI Cookbook, and verified that my content offers new insights or unique information compared to existing documentation.
  - [ ] Spelling and Grammar: I have checked for spelling or grammatical mistakes.
  - [ ] Clarity: I have done a final read-through and verified that my submission is well-organized and easy to understand.
  - [ ] Correctness: The information I include is correct and all of my code executes successfully.
  - [ ] Completeness: I have explained everything fully, including all necessary references and citations.

We will rate each of these areas on a scale from 1 to 4, and will only accept contributions that score 3 or higher on all areas. Refer to our [contribution guidelines](https://github.com/openai/openai-cookbook/blob/main/CONTRIBUTING.md) for more details.

### CODE REVIEW

**Comment by royziv11** [comment]
@shyamal-anadkat - Adding you as a reviewer to this PR. I went through and updated the cookbook and added in a section on parallel function calling. Would appreciate some feedback here on how to ensure it remains up to standard. Thank you!

**Review by simonpfish — APPROVED** [review]
This is great! Much needed update. 

It could also be good to show how this can be done with the Python SDK rather than using `requests`, but perhaps that can be a follow-up PR

**Comment by simonpfish** [comment]
@shyamal-anadkat I'll leave this open for a bit in case you'd like to take a look as well

File: examples/How_to_call_functions_with_chat_models.ipynb [hunk 1]
```diff
@@ -8,11 +8,11 @@
    "source": [
     "# How to call functions with chat models\n",
     "\n",
-    "This notebook covers how to use the Chat Completions API in combination with external functions to extend the capabilities of GPT models.\n",
+    "This notebook covers how to use the Chat Completions API and Assistants API in combination with external functions to extend the capabilities of GPT models.\n",
```
Comment (simonpfish) [inline examples/How_to_call_functions_with_chat_models.ipynb:11]:
```suggestion
    "This notebook covers how to use the Chat Completions API in combination with external functions to extend the capabilities of GPT models.\n",
```

It looks like we don't actually cover the assistants API here?

File: examples/How_to_call_functions_with_chat_models.ipynb [hunk 2]
```diff
@@ -52,7 +52,7 @@
   },
   {
    "cell_type": "code",
-   "execution_count": 1,
+   "execution_count": 264,
```
Comment (simonpfish) [inline examples/How_to_call_functions_with_chat_models.ipynb:55]: Let's restart the kernel and rerun from scratch to clean up and reset these execution counts

File: examples/How_to_call_functions_with_chat_models.ipynb [hunk 3]
```diff
@@ -467,11 +482,56 @@
...
     "chat_response.json()[\"choices\"][0][\"message\"]\n"
    ]
   },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Parallel Function Calling\n",
+    "\n",
+    "Newer models like gpt-4-1106-preview or gpt-3.5-turbo-1106 can call multiple functions in one turn"
```
Comment (simonpfish) [inline examples/How_to_call_functions_with_chat_models.ipynb:496]:
```suggestion
    "Newer models like gpt-4-1106-preview or gpt-3.5-turbo-1106 can call multiple functions in one turn."
```

File: examples/How_to_call_functions_with_chat_models.ipynb [hunk 4]
```diff
@@ -79,22 +79,22 @@
   },
   {
    "cell_type": "code",
-   "execution_count": 2,
+   "execution_count": 266,
```
Comment (shyamal-anadkat) [inline examples/How_to_call_functions_with_chat_models.ipynb:82]: nit: refresh

File: examples/How_to_call_functions_with_chat_models.ipynb [hunk 5]
```diff
@@ -79,22 +79,22 @@
...
-    "        json_data.update({\"functions\": functions})\n",
-    "    if function_call is not None:\n",
-    "        json_data.update({\"function_call\": function_call})\n",
+    "    if tools is not None:\n",
+    "        json_data.update({\"tools\": tools})\n",
+    "    if tool_choice is not None:\n",
+    "        json_data.update({\"tool_choice\": tool_choice})\n",
     "    try:\n",
     "        response = requests.post(\n",
     "            \"https://api.openai.com/v1/chat/completions\",\n",
```
Comment (shyamal-anadkat) [inline examples/How_to_call_functions_with_chat_models.ipynb:100]: if you have some time, would refactor these to use the sdk

File: examples/How_to_call_functions_with_chat_models.ipynb [hunk 6]
```diff
@@ -696,34 +760,40 @@
...
-    "if assistant_message.get(\"function_call\"):\n",
+    "if assistant_message.get(\"tool_calls\"):\n",
     "    results = execute_function_call(assistant_message)\n",
-    "    messages.append({\"role\": \"function\", \"content\": results, \"name\": assistant_message[\"function_call\"][\"name\"]})\n",
+    "    messages.append({\"role\": \"tool\", \"tool_call_id\": assistant_message[\"tool_calls\"][0]['id'], \"name\": assistant_message[\"tool_calls\"][0][\"function\"][\"name\"], \"content\": results})\n",
     "pretty_print_conversation(messages)"
    ]
   },
+  {
+   "cell_type": "markdown",
```
Comment (shyamal-anadkat) [inline examples/How_to_call_functions_with_chat_models.ipynb:793]: remove?

**Review by shyamal-anadkat — COMMENTED** [review]
left few comments. lgtm overall, thanks

**Comment by royziv11** [comment]
@simonpfish - I cleaned up the execution count and will update to use the python SDK in a separate PR. LMK if this is good to merge and I'll get started on the other PR.

**Review by shyamal-anadkat — APPROVED** [review]

**Comment by sagarspatil** [comment]
> @simonpfish - I cleaned up the execution count and will update to use the python SDK in a separate PR. LMK if this is good to merge and I'll get started on the other PR.

Hey. Any ETA on when you will have this updated to use python SDK instead of requests?
Thanks!

**Pull request merged**